### DISCUSSION TITLE
What is the best way to present alert with custom view?

### DISCUSSION THREAD

**Discussion by havilog** [pointfreeco/swift-composable-architecture#2007] [Q&A] [answered]
I have my own existing Design System. 

so I have to present the alert with specific design and want to migrate it to TCA.

but `AlertState` can't cover my custom views.

so I'm wondering if I could present a custom view, not AlertState, when presenting an alert.

Is it impossible? or Is it being prepared? or shoud I custom AlertState?

**Comment by mbrandonw** [comment, marked as answer]
Hi @havilog, we have an example of this in our [isowords](https://github.com/pointfreeco/isowords/) project. There's a custom UI component we have called `BottomMenu`, and it was first designed in plain, vanilla SwiftUI. This means it's driven off a [binding](https://github.com/pointfreeco/isowords/blob/455238749625d8a84c23ee641d537971695416fc/Sources/BottomMenu/BottomMenu.swift#L72) and comes with a [custom datatype](https://github.com/pointfreeco/isowords/blob/455238749625d8a84c23ee641d537971695416fc/Sources/BottomMenu/BottomMenu.swift#L4-L8) that describes how to present a bottom menu.

_Then_ we made a [TCA friendly interface](https://github.com/pointfreeco/isowords/blob/455238749625d8a84c23ee641d537971695416fc/Sources/BottomMenu/ComposableBottomMenu.swift) for the bottom menu where it takes a `Store` of `BottomMenu` domain and uses `viewStore.binding` to derive a binding to hand to the vanilla SwiftUI component.

Does that help?

**Reply by havilog** [reply]
Hi @mbrandonw, I've checked isoword `BottomMenu` and it helped me thanks.
but I'm curious that is there any way to adapt nicely using `@PresentableState` and custom view?

**Reply by junyng** [reply]
I'm also curious about the above question..
Is there any way to adapt `@PresentationState` to vanilla SwiftUI component?
I've checked other component like `.sheet` but I can't use `PresentationStore` because It was used `@_spi` annotation

**Reply by stephencelis** [reply]
@junyng You are free to `@_spi(Presentation) import ComposableArchitecture` to access these APIs. They should be relatively stable, but the `@_spi` will give us a little more freedom to change/evolve them before making them totally public.

**Reply by junyng** [reply]
@stephencelis
oh, i was wrong.. sorry
According to your description, it's possible to apply `PresentationStore` for custom components.
I'll try it thanks.
